File: server/app/models/transaction.py
from sqlalchemy import Column, String, Float, Integer, DateTime, ForeignKey, Text, Boolean
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from datetime import datetime

from app.db.base import Base


class Transaction(Base):
    """
    Модель транзакции (доходы/расходы)
    Хранит историю финансовых операций пользователя
    """
    __tablename__ = "transactions"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    category_id = Column(Integer, ForeignKey("categories.id", ondelete="SET NULL"), nullable=True, index=True)

    # Основные данные
    amount = Column(Float, nullable=False)  # Положительное = доход, отрицательное = расход
    description = Column(Text, nullable=True)
    date = Column(DateTime, default=datetime.utcnow, nullable=False, index=True)

    # Данные чека (опционально)
    receipt_data = Column(JSONB, nullable=True)  # JSON с данными чека: QR, items, retailer

    # ML метаданные
    ml_category = Column(String(100), nullable=True)  # Категория от ML модели
    ml_confidence = Column(Float, nullable=True)  # Уверенность ML модели (0-1)
    is_anomaly = Column(Boolean, default=False, nullable=False)  # Флаг аномальной транзакции

    # Метаданные синхронизации
    device_id = Column(String(100), nullable=True)  # ID устройства (для синхронизации)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    version = Column(Integer, default=1, nullable=False)  # Версия для разрешения конфликтов

    # Relationships
    user = relationship("User", back_populates="transactions")
    category = relationship("Category", back_populates="transactions")

    @property
    def is_expense(self) -> bool:
        """Является ли транзакция расходом"""
        return self.amount < 0

    @property
    def absolute_amount(self) -> float:
        """Абсолютное значение суммы"""
        return abs(self.amount)

    def __repr__(self):
        return f"<Transaction(id={self.id}, amount={self.amount}, category_id={self.category_id})>"

File: lib/models/transaction.dart
import 'package:hive/hive.dart';

part 'transaction.g.dart';

@HiveType(typeId: 1)
class Transaction extends HiveObject {
  @HiveField(0)
  String id; // UUID

  @HiveField(1)
  double amount; // Положительное = доход, отрицательное = расход

  @HiveField(2)
  int categoryId;

  @HiveField(3)
  DateTime date; // Всегда с временем 00:00:00

  @HiveField(4)
  String description;

  @HiveField(5)
  Map<String, dynamic>? receiptData; // Данные чека (опционально)

  Transaction({
    required this.id,
    required this.amount,
    required this.categoryId,
    required DateTime date,
    required this.description,
    this.receiptData,
  }) : date = DateTime(date.year, date.month, date.day); // Обнуляем время!

  // Проверка, является ли транзакция доходом
  bool get isIncome => amount > 0;

  // Проверка, является ли транзакция расходом
  bool get isExpense => amount < 0;

  // Абсолютное значение суммы (для отображения)
  double get absoluteAmount => amount.abs();
}
